File: 2021/python/day02.py
def Run(filepath):
    with open(filepath) as input:
        directions = list(map(str, input.read().split('\n')))

    print('Day 2')
    Part1(directions)
    Part2(directions)
    
def Part1(directions):
    horizontal = 0
    depth = 0

    for instruction in directions:
        #Split the full instruction into direction and distance
        direction = instruction.split(' ')[0]
        distance = instruction.split(' ')[1]
        if direction == 'forward':
            horizontal += int(distance)
        elif direction == 'down':
            depth += int(distance)
        elif direction == 'up':
            depth -= int(distance)

    answer = horizontal * depth

    print('Part 1: ' + str(answer))
    #The result should be 1989014

def Part2(directions):    
    horizontal = 0
    depth = 0
    aim = 0

    for instruction in directions:
        #Split the full instruction into direction and distance
        direction = instruction.split(' ')[0]
        distance = instruction.split(' ')[1]
        if direction == 'forward':
            horizontal += int(distance)
            depth += aim * int(distance)
        elif direction == 'down':
            aim += int(distance)
        elif direction == 'up':
            aim -= int(distance)

    answer = horizontal * depth

    print('Part 2: ' + str(answer))
    #Result should be 2006917119

File: 2021/csharp/app/Day02.cs
namespace AdventOfCode;
public class Day02 : Day{
    public override Int64 Part1(string filepath){
        string[] file = File.ReadAllLines(filepath);
        List<string> directions = new List<string>(file);
        int destination = PlotCourse(directions);
        
        System.Console.WriteLine($"Day 2-1: {destination}");
        return destination;
    }
    public override Int64 Part2(string filepath){
        string[] file = File.ReadAllLines(filepath);
        List<string> directions = new List<string>(file);
        int destination = PlotCourse(directions, 0);

        System.Console.WriteLine($"Day 2-2: {destination}");
        return destination;
    }
    static int PlotCourse(List<string> directions){
        //For part 1, where the course just follows the directions
        int horizontal = 0, depth = 0;
        foreach (string direction in directions){
            var (heading, distance) = SplitInstruction(direction);
            switch (heading){
                case "forward":
                    horizontal += distance;
                    break;
                case "down":
                    depth += distance;
                    break;
                case "up":
                    depth -= distance;
                    break;
            }
        }
        int destination = horizontal * depth;
        return destination;
    }
    static int PlotCourse(List<string> directions, int aim){
        //For part 2, where aim affects how directions are followed
        int horizontal = 0, depth = 0;
        foreach (string direction in directions){
            var (heading, distance) = SplitInstruction(direction);
            switch (heading){
                case "forward":
                    horizontal += distance;
                    depth += aim * distance;
                    break;
                case "down":
                    aim += distance;
                    break;
                case "up":
                    aim -= distance;
                    break;
            }
        }
        int destination = horizontal * depth;
        return destination;
    }
    static (string heading, int distance) SplitInstruction(string direction){
        //Split the full instruction into a tuple containing direction and distance
        string[] directionSplit = direction.Split(" ");
        (string, int) directionParts = (directionSplit[0], Convert.ToInt32(directionSplit[1]));
        return directionParts;
    }
}